Add 26.06 migration guide (#2344)
rapidsai/rmm: docs/migration_guide_2606.md

# RMM 26.06 Migration Guide

RMM 26.06 completes the migration to CCCL-native memory resources. This release
removes `device_memory_resource`, eliminates the `Upstream` template parameter
from all resources and adaptors, introduces `cuda::mr::shared_resource`-based
ownership, and removes several legacy APIs. This guide covers the breaking
changes and shows how to update consuming code.

## Summary of Breaking Changes

RMM's memory resource model changes from inheritance-based resources to

> CCCL-native value resources:

Reviewer: Please link to the CCCL docs.

- `device_memory_resource` is removed; resources must now satisfy CCCL's
  `cuda::mr::resource` concept directly.
- Resources and adaptors are no longer templated on `Upstream`.
- Upstream resources are stored as
  `cuda::mr::any_resource<cuda::mr::device_accessible>` instead of raw pointers.
- Resources with non-trivial state use `cuda::mr::shared_resource<Impl>` for
  internal shared ownership.

Several APIs change as a consequence:

- Allocation order changes from `allocate(size, stream)` to
  `allocate(stream, size, alignment)`. This aligns with the CCCL memory
  resource design.
- Pointer-based per-device resource setters are replaced by `any_resource`
  setters.
- `owning_wrapper` is removed.
- Resource implementations are compiled into `librmm` instead of being fully

>   header-only.

Reviewer: This migration guide mixes changes that _implementers_ of memory resources need to care about from changes that _users_ of memory resources need to care about.

I think we should separate those, bring the "user" case to the fore, and then cover the "implement" case later.

These types are implicitly constructible from any resource satisfying the
corresponding CCCL concept, and from `cuda::mr::any_resource`.

Since `device_memory_resource` no longer exists, `device_async_resource_ref`
can no longer be constructed from a `device_memory_resource*`. Code that
constructed a `resource_ref` from a `device_memory_resource` pointer must
be updated to pass the concrete resource type directly.

(rmm-2604-2606-compiled-resources)=
### Resource Implementations Are Compiled, Not Header-Only

In 26.04, RMM resource implementations were entirely header-only (template
definitions lived in the headers). In 26.06, many resources and adaptors have
their implementations in compiled translation units under `cpp/src/mr/`. The
public headers contain only the class declaration and inline accessors.

This means downstream projects that use RMM memory resources **must link against
`librmm`**. If your project previously relied on RMM being header-only for the
memory resource layer, you will need to update your build system to link the
`rmm` library target.

For CMake users, this should already work if you use `find_package(rmm)` or

> `rapids_find_package(rmm)` — the imported `rmm::rmm` target handles linking.

Reviewer: ```suggestion
For CMake users, this should already work if you use `find_package(rmm)` or
`rapids_find_package(rmm)`; the imported `rmm::rmm` target handles linking.
```

## Memory Resource Changes

### C++

(rmm-2604-2606-device-memory-resource)=
### `device_memory_resource` Base Class Is Removed

**Header removed:** `rmm/mr/device_memory_resource.hpp`

In 26.04, all RMM memory resources — both base resources like

> `cuda_memory_resource` and adaptors like `pool_memory_resource` — inherited from

Reviewer: ```suggestion
In 26.04, all RMM memory resources inherited from
```

`rmm::mr::device_memory_resource` and overrode `do_allocate` / `do_deallocate`
virtual methods. In 26.06, this base class no longer exists. Every resource

> (base and adaptor alike) instead satisfies CCCL's `cuda::mr::resource` concept

Reviewer: ```suggestion
instead satisfies CCCL's `cuda::mr::resource` concept
```

  `deallocate` with the CCCL signature instead (see
  {ref}`Allocation Signature Changes <rmm-2604-2606-allocation-signatures>`).
- Remove calls to `device_memory_resource::allocate(bytes, stream)`. Use
  `device_async_resource_ref` or call the resource's `allocate(stream, bytes)` directly.
- Replace `device_memory_resource*` variables with `device_async_resource_ref`
  (non-owning) or `cuda::mr::any_resource<cuda::mr::device_accessible>` (owning).

```cpp
// 26.04
#include <rmm/mr/device_memory_resource.hpp>

class my_resource : public rmm::mr::device_memory_resource {
  void* do_allocate(std::size_t bytes, rmm::cuda_stream_view stream) override { ... }
  void do_deallocate(void* ptr, std::size_t bytes, rmm::cuda_stream_view stream) override { ... }
};

// 26.06
#include <cuda/memory_resource>

class my_resource {
 public:
  void* allocate(cuda::stream_ref stream, std::size_t bytes, std::size_t alignment = 256) { ... }
  void deallocate(cuda::stream_ref stream, void* ptr, std::size_t bytes,
                  std::size_t alignment = 256) noexcept { ... }
  bool operator==(my_resource const&) const noexcept { ... }

>   bool operator!=(my_resource const& other) const noexcept { return !(*this == other); }

Reviewer: Is this required or does c++20 put it in automatically for us if we define operator==?

Author: We only require C++17 so there is no assumption that `operator!=` is synthesized yet. See #2410.

- **`deallocate` is `noexcept`**.

Resources also provide synchronous methods `allocate_sync(bytes, alignment)` and
`deallocate_sync(ptr, bytes, alignment)`.

If you call allocation methods through `device_async_resource_ref`, the
wrapper preserves the `allocate(stream, bytes)` calling convention.

### Resources Use `cuda::mr::shared_resource` for Ownership

All RMM resources with non-trivial state now inherit from
`cuda::mr::shared_resource<detail::*_impl>`. This includes pool resources,
all adaptors, and also base resources like `cuda_async_memory_resource` and
`callback_memory_resource`. Only truly trivial resources like
`cuda_memory_resource` remain standalone. This means:

- **Resources are copyable** and copies share ownership of the underlying state.
- Resource lifetime is managed by reference counting internally; no need
  for `std::shared_ptr` or `std::unique_ptr` wrappers externally.
- Each resource class is a thin public API shell; implementation details
  live in a `detail::*_impl` class.

```cpp
// 26.06 — copying a pool shares the same underlying pool
rmm::mr::pool_memory_resource pool{cuda_mr, pool_size};
auto pool_copy = pool;  // both reference the same pool state

> ```

Reviewer: question: If I am migrating as a user, do I care about the implementation? I think no, only that I don't need smart pointer management?
